File: SLR_Assignment_Churn_out_rate.py
# -*- coding: utf-8 -*-
"""
Created on Tue Apr 28 16:02:39 2020

@author: RAVI
"""
# For reading data set
# importing necessary libraries
import pandas as pd # deals with data frame  
import numpy as np  # deals with numerical values

SH_CO = pd.read_csv("C:/RAVI/Data science/Assignments/Module 6 Simple linear regression/DataSets/emp_data.csv")

import matplotlib.pylab as plt #for different types of plots
SH_CO.Salary_hike
plt.scatter(x=SH_CO['Salary_hike'], y=SH_CO['Churn_out_rate'],color='green')# Scatter plot

np.corrcoef(SH_CO.Salary_hike, SH_CO.Churn_out_rate) #correlation

help(np.corrcoef)

import statsmodels.formula.api as smf
plt.hist(SH_CO["Salary_hike"])
plt.hist(SH_CO["Churn_out_rate"])
model = smf.ols('Churn_out_rate ~ Salary_hike', data=SH_CO).fit()
model.summary()

#values prediction
#Confidence interval Calculation
pred1 = model.predict(pd.DataFrame(SH_CO['Salary_hike']))
pred1
print (model.conf_int(0.95)) # 95% confidence interval

res = SH_CO.Churn_out_rate - pred1
sqres = res*res
mse = np.mean(sqres)
rmse = np.sqrt(mse)

######### Model building on Transformed Data#############

# Log Transformation
# x = log(Salary_hike); y = Churn_out_rate
plt.scatter(x=np.log(SH_CO['Salary_hike']),y=SH_CO['Churn_out_rate'],color='brown')
np.corrcoef(np.log(SH_CO.Salary_hike), SH_CO.Churn_out_rate) #correlation

model2 = smf.ols('Churn_out_rate ~ np.log(Salary_hike)',data=SH_CO).fit()
model2.summary()

pred2 = model2.predict(pd.DataFrame(SH_CO['Salary_hike']))
pred2
print(model2.conf_int(0.95)) # 95% confidence level

res2 = SH_CO.Churn_out_rate - pred2
sqres2 = res2*res2
mse2 = np.mean(sqres2)
rmse2 = np.sqrt(mse2)

############Polynomial model with 2 degree (quadratic model)  ;x = Salary_hike*Salary_hike; y = Churn_out_rate############
#### input=x & X^2 (2-degree); output=y  ####
model4 = smf.ols('Churn_out_rate ~ Salary_hike+I(Salary_hike*Salary_hike)', data=SH_CO).fit()
model4.summary()

pred_p2 = model4.predict(pd.DataFrame(SH_CO['Salary_hike']))
pred_p2

print(model4.conf_int(0.95)) # 95% confidence level

res4 = SH_CO.Churn_out_rate - pred_p2
sqres4 = res4*res4
mse4 = np.mean(sqres4)
rmse4 = np.sqrt(mse4)

###########Polynomial model with 3 degree (quadratic model)  ;x = Salary_hike*Salary_hike*Salary_hike; y = Churn_out_rate############
#### input=x & X^2 (2-degree); output=y  ####
model5 = smf.ols('Churn_out_rate ~ Salary_hike+I(Salary_hike*Salary_hike)+I(Salary_hike*Salary_hike*Salary_hike)', data=SH_CO).fit()
model5.summary()

pred_p3 = model5.predict(pd.DataFrame(SH_CO['Salary_hike']))
pred_p3

print(model5.conf_int(0.95)) # 95% confidence level

res5 = SH_CO.Churn_out_rate - pred_p3
sqres5 = res5*res5
mse5 = np.mean(sqres5)
rmse5 = np.sqrt(mse5)

File: SLR_Assignment_Churn_out_rate.R
#  Simple Linear Regression Assignment #
#  3) Emp_data -> Build a prediction model for Churn_out_rate 
#  Do the necessary transformations for input variables for getting better R^2 value for the model prepared.

library(readr)
library(ggplot2)
library(moments)
SH_CO <- read_csv("C:/RAVI/Data science/Assignments/Module 6 Simple linear regression/DataSets/emp_data.csv")
View(SH_CO)
attach(SH_CO)
summary(SH_CO)
range(SH_CO$Salary_hike)
range(SH_CO$Churn_out_rate)
skewness(SH_CO$Salary_hike)
skewness(SH_CO$Churn_out_rate)

#Exploratory Data Analysis
boxplot(SH_CO$Salary_hike)
boxplot(SH_CO$Churn_out_rate)

#scatter plot for Caloriesconsumed vs Weightgained (Plot x,y)
plot(Salary_hike,Churn_out_rate)

#calculate correlation coefficient
cor(Salary_hike,Churn_out_rate)

#Simple Regression model
reg <- lm(Churn_out_rate~Salary_hike,data =SH_CO)
summary(reg)

#values prediction
#Confidence interval Calculation
confint(reg,level = 0.95)
pred <- predict(reg,interval = "predict")
#predict function gives fit value and its lower and upeer values as a range
pred <- as.data.frame(pred)
pred

#####Plot Graph for both Actual values and also the predicted linear Graph(Actual:Red,Predicted:Blue)#########
ggplot() + 
  geom_point(aes(x =`Salary_hike` , y =`Churn_out_rate` ),
             colour='red') + 
  geom_line(aes(x = `Salary_hike`, y = predict(reg, newdata=SH_CO)),
            colour='blue') + 
  ggtitle('Salary_hike vs Churn_out_rate') +
  xlab('Salary_hike') +
  ylab('Churn_out_rate')

cor(pred$fit,`Churn_out_rate`)

#Calculate Residuals "Errors"
reg$residuals
reg$residuals^2
mean(reg$residuals^2)
rmse <- sqrt(mean(reg$residuals^2))
rmse

############ Applying transformations##############
############ lOGORITHMIC MODEL    x = log(Salary_hike); y = Churn_out_rate ############
plot(log(`Salary_hike`),`Churn_out_rate`)
cor(log(`Salary_hike`),`Churn_out_rate`)

log_reg <- lm(`Churn_out_rate` ~ log(`Salary_hike`),data = SH_CO)
summary(log_reg)

#values prediction
#Confidence interval Calculation
confint(log_reg,level = 0.95)
pred_log <- predict(log_reg,interval ="predict")
#predict function gives fit value and its lower and upeer values as a range
pred_log <- as.data.frame(pred_log)
pred_log

rmse_log <- sqrt(mean(log_reg$residuals^2)) 
rmse_log
######or####
res_log=`Churn_out_rate`-pred_log$fit
rmse_log<-sqrt(mean(res_log^2))
rmse_log
##########Plot Graph for both Actual values and also the predicted linear Graph(Actual:Red,Predicted:Blue)#########
ggplot() + 
  geom_point(aes(x =`Salary_hike` , y =`Churn_out_rate` ),
             colour='red') + 
  geom_line(aes(x =`Salary_hike`, y = predict(log_reg, newdata=SH_CO)),
            colour='blue') + 
  ggtitle('Salary_hike vs Churn_out_rate') +
  xlab('Salary_hike') +
  ylab('Churn_out_rate')


############Polynomial model with 2 degree (quadratic model)  ;x =Salary_hike^2 ; y = Churn_out_rate ############
#### input=x & X^2 (2-degree); output=y  ####
reg_quad2<- lm(`Churn_out_rate` ~ `Salary_hike`+I(`Salary_hike`*`Salary_hike`),data = SH_CO)
summary(reg_quad2)

#prediction
#Confidence interval Calculation
confint(reg_quad2,level = 0.95)
pred_quad2<-predict(reg_quad2,interval = "predict")
pred_quad2  <- as.data.frame(pred_quad2)
pred_quad2

resq=`Churn_out_rate`-pred_quad2$fit
rmse_quad<-sqrt(mean(resq^2))
rmse_quad


##########Plot Graph for both Actual values and also the predicted linear Graph(Actual:Red,Predicted:Blue)#########
ggplot() + 
  geom_point(aes(x =`Salary_hike` , y =`Churn_out_rate` ),
             colour='red') + 
  geom_line(aes(x = `Salary_hike`, y = predict(reg_quad2, newdata=SH_CO)),
            colour='blue') + 
  ggtitle('Salary_hike vs Churn_out_rate') +
  xlab('Salary_hike') +
  ylab('Churn_out_rate')

############Polynomial model with 3 degree (quadratic model)  ;x = Salary_hike^3; y = Churn_out_rate ############
#### input=x & X^2 & x^3 (3-degree); output=y  ####
reg_quad3<- lm(`Churn_out_rate` ~ `Salary_hike`+I(`Salary_hike`*`Salary_hike`)+I(`Salary_hike`*`Salary_hike`*`Salary_hike`),data = SH_CO)
summary(reg_quad3)

#prediction
#Confidence interval Calculation
confint(reg_quad3,level = 0.95)
pred_quad3<-predict(reg_quad3,interval = "predict")
pred_quad3  <- as.data.frame(pred_quad3)
pred_quad3

resq3=`Churn_out_rate`-pred_quad3$fit
rmse_quad3<-sqrt(mean(resq3^2))
rmse_quad3
##########Plot Graph for both Actual values and also the predicted linear Graph(Actual:Red,Predicted:Blue)#########
ggplot() + 
  geom_point(aes(x =`Salary_hike` , y =`Churn_out_rate` ),
             colour='red') + 
  geom_line(aes(x = `Salary_hike`, y = predict(reg_quad3, newdata= SH_CO)),
            colour='blue') + 
  ggtitle('Salary_hike vs Churn_out_rate') +
  xlab('Salary_hike') +
  ylab('Churn_out_rate')



